File: nestrefresh/src/main/java/com/todou/nestrefresh/material/resources/NRTintTypedArray.java
package com.todou.nestrefresh.material.resources;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.v7.content.res.AppCompatResources;
import android.util.AttributeSet;
import android.util.TypedValue;

public class NRTintTypedArray {
    private final Context mContext;
    private final TypedArray mWrapped;
    private TypedValue mTypedValue;

    public static NRTintTypedArray obtainStyledAttributes(Context context, AttributeSet set, int[] attrs) {
        return new NRTintTypedArray(context, context.obtainStyledAttributes(set, attrs));
    }

    public static NRTintTypedArray obtainStyledAttributes(Context context, AttributeSet set, int[] attrs, int defStyleAttr, int defStyleRes) {
        return new NRTintTypedArray(context, context.obtainStyledAttributes(set, attrs, defStyleAttr, defStyleRes));
    }

    public static NRTintTypedArray obtainStyledAttributes(Context context, int resid, int[] attrs) {
        return new NRTintTypedArray(context, context.obtainStyledAttributes(resid, attrs));
    }

    private NRTintTypedArray(Context context, TypedArray array) {
        this.mContext = context;
        this.mWrapped = array;
    }

    public Drawable getDrawable(int index) {
        if (this.mWrapped.hasValue(index)) {
            int resourceId = this.mWrapped.getResourceId(index, 0);
            if (resourceId != 0) {
                return AppCompatResources.getDrawable(this.mContext, resourceId);
            }
        }

        return this.mWrapped.getDrawable(index);
    }

    public int length() {
        return this.mWrapped.length();
    }

    public int getIndexCount() {
        return this.mWrapped.getIndexCount();
    }

    public int getIndex(int at) {
        return this.mWrapped.getIndex(at);
    }

    public Resources getResources() {
        return this.mWrapped.getResources();
    }

    public CharSequence getText(int index) {
        return this.mWrapped.getText(index);
    }

    public String getString(int index) {
        return this.mWrapped.getString(index);
    }

    public String getNonResourceString(int index) {
        return this.mWrapped.getNonResourceString(index);
    }

    public boolean getBoolean(int index, boolean defValue) {
        return this.mWrapped.getBoolean(index, defValue);
    }

    public int getInt(int index, int defValue) {
        return this.mWrapped.getInt(index, defValue);
    }

    public float getFloat(int index, float defValue) {
        return this.mWrapped.getFloat(index, defValue);
    }

    public int getColor(int index, int defValue) {
        return this.mWrapped.getColor(index, defValue);
    }

    public ColorStateList getColorStateList(int index) {
        if (this.mWrapped.hasValue(index)) {
            int resourceId = this.mWrapped.getResourceId(index, 0);
            if (resourceId != 0) {
                ColorStateList value = AppCompatResources.getColorStateList(this.mContext, resourceId);
                if (value != null) {
                    return value;
                }
            }
        }

        return this.mWrapped.getColorStateList(index);
    }

    public int getInteger(int index, int defValue) {
        return this.mWrapped.getInteger(index, defValue);
    }

    public float getDimension(int index, float defValue) {
        return this.mWrapped.getDimension(index, defValue);
    }

    public int getDimensionPixelOffset(int index, int defValue) {
        return this.mWrapped.getDimensionPixelOffset(index, defValue);
    }

    public int getDimensionPixelSize(int index, int defValue) {
        return this.mWrapped.getDimensionPixelSize(index, defValue);
    }

    public int getLayoutDimension(int index, String name) {
        return this.mWrapped.getLayoutDimension(index, name);
    }

    public int getLayoutDimension(int index, int defValue) {
        return this.mWrapped.getLayoutDimension(index, defValue);
    }

    public float getFraction(int index, int base, int pbase, float defValue) {
        return this.mWrapped.getFraction(index, base, pbase, defValue);
    }

    public int getResourceId(int index, int defValue) {
        return this.mWrapped.getResourceId(index, defValue);
    }

    public CharSequence[] getTextArray(int index) {
        return this.mWrapped.getTextArray(index);
    }

    public boolean getValue(int index, TypedValue outValue) {
        return this.mWrapped.getValue(index, outValue);
    }

    public int getType(int index) {
        if (Build.VERSION.SDK_INT >= 21) {
            return this.mWrapped.getType(index);
        } else {
            if (this.mTypedValue == null) {
                this.mTypedValue = new TypedValue();
            }

            this.mWrapped.getValue(index, this.mTypedValue);
            return this.mTypedValue.type;
        }
    }

    public boolean hasValue(int index) {
        return this.mWrapped.hasValue(index);
    }

    public TypedValue peekValue(int index) {
        return this.mWrapped.peekValue(index);
    }

    public String getPositionDescription() {
        return this.mWrapped.getPositionDescription();
    }

    public void recycle() {
        this.mWrapped.recycle();
    }

    @RequiresApi(21)
    public int getChangingConfigurations() {
        return this.mWrapped.getChangingConfigurations();
    }
}

File: nestrefresh/src/main/java/com/todou/nestrefresh/material/resources/NRTintTypedArray.kt
package com.todou.nestrefresh.material.resources

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.content.res.TypedArray
import android.graphics.drawable.Drawable
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.appcompat.content.res.AppCompatResources
import android.util.AttributeSet
import android.util.TypedValue

class NRTintTypedArray private constructor(private val mContext: Context, private val mWrapped: TypedArray) {
    private var mTypedValue: TypedValue? = null

    val indexCount: Int
        get() = this.mWrapped.indexCount

    val resources: Resources
        get() = this.mWrapped.resources

    val positionDescription: String
        get() = this.mWrapped.positionDescription

    val changingConfigurations: Int
        @RequiresApi(21)
        get() = this.mWrapped.changingConfigurations

    fun getDrawable(index: Int): Drawable? {
        if (this.mWrapped.hasValue(index)) {
            val resourceId = this.mWrapped.getResourceId(index, 0)
            if (resourceId != 0) {
                return AppCompatResources.getDrawable(this.mContext, resourceId)
            }
        }

        return this.mWrapped.getDrawable(index)
    }

    fun length(): Int {
        return this.mWrapped.length()
    }

    fun getIndex(at: Int): Int {
        return this.mWrapped.getIndex(at)
    }

    fun getText(index: Int): CharSequence {
        return this.mWrapped.getText(index)
    }

    fun getString(index: Int): String? {
        return this.mWrapped.getString(index)
    }

    fun getNonResourceString(index: Int): String {
        return this.mWrapped.getNonResourceString(index)
    }

    fun getBoolean(index: Int, defValue: Boolean): Boolean {
        return this.mWrapped.getBoolean(index, defValue)
    }

    fun getInt(index: Int, defValue: Int): Int {
        return this.mWrapped.getInt(index, defValue)
    }

    fun getFloat(index: Int, defValue: Float): Float {
        return this.mWrapped.getFloat(index, defValue)
    }

    fun getColor(index: Int, defValue: Int): Int {
        return this.mWrapped.getColor(index, defValue)
    }

    fun getColorStateList(index: Int): ColorStateList? {
        if (this.mWrapped.hasValue(index)) {
            val resourceId = this.mWrapped.getResourceId(index, 0)
            if (resourceId != 0) {
                val value = AppCompatResources.getColorStateList(this.mContext, resourceId)
                if (value != null) {
                    return value
                }
            }
        }

        return this.mWrapped.getColorStateList(index)
    }

    fun getInteger(index: Int, defValue: Int): Int {
        return this.mWrapped.getInteger(index, defValue)
    }

    fun getDimension(index: Int, defValue: Float): Float {
        return this.mWrapped.getDimension(index, defValue)
    }

    fun getDimensionPixelOffset(index: Int, defValue: Int): Int {
        return this.mWrapped.getDimensionPixelOffset(index, defValue)
    }

    fun getDimensionPixelSize(index: Int, defValue: Int): Int {
        return this.mWrapped.getDimensionPixelSize(index, defValue)
    }

    fun getLayoutDimension(index: Int, name: String): Int {
        return this.mWrapped.getLayoutDimension(index, name)
    }

    fun getLayoutDimension(index: Int, defValue: Int): Int {
        return this.mWrapped.getLayoutDimension(index, defValue)
    }

    fun getFraction(index: Int, base: Int, pbase: Int, defValue: Float): Float {
        return this.mWrapped.getFraction(index, base, pbase, defValue)
    }

    fun getResourceId(index: Int, defValue: Int): Int {
        return this.mWrapped.getResourceId(index, defValue)
    }

    fun getTextArray(index: Int): Array<CharSequence> {
        return this.mWrapped.getTextArray(index)
    }

    fun getValue(index: Int, outValue: TypedValue): Boolean {
        return this.mWrapped.getValue(index, outValue)
    }

    fun getType(index: Int): Int {
        if (Build.VERSION.SDK_INT >= 21) {
            return this.mWrapped.getType(index)
        } else {
            if (this.mTypedValue == null) {
                this.mTypedValue = TypedValue()
            }

            this.mWrapped.getValue(index, this.mTypedValue)
            return this.mTypedValue!!.type
        }
    }

    fun hasValue(index: Int): Boolean {
        return this.mWrapped.hasValue(index)
    }

    fun peekValue(index: Int): TypedValue {
        return this.mWrapped.peekValue(index)
    }

    fun recycle() {
        this.mWrapped.recycle()
    }

    companion object {

        fun obtainStyledAttributes(context: Context, set: AttributeSet, attrs: IntArray): NRTintTypedArray {
            return NRTintTypedArray(context, context.obtainStyledAttributes(set, attrs))
        }

        fun obtainStyledAttributes(
            context: Context,
            set: AttributeSet,
            attrs: IntArray,
            defStyleAttr: Int,
            defStyleRes: Int
        ): NRTintTypedArray {
            return NRTintTypedArray(context, context.obtainStyledAttributes(set, attrs, defStyleAttr, defStyleRes))
        }

        fun obtainStyledAttributes(context: Context, resid: Int, attrs: IntArray): NRTintTypedArray {
            return NRTintTypedArray(context, context.obtainStyledAttributes(resid, attrs))
        }
    }
}